jgiven-core/src/main/java/com/tngtech/jgiven/report/asciidoc/AsciiDocReportModelVisitor.java

```diff
@@ -0,0 +1,135 @@
+    public void visit(final ReportModel reportModel) {
+        String featureName = Optional.ofNullable(reportModel.getName())
+            .filter(name -> !name.isEmpty())
+            .orElse(reportModel.getClassName());
+
+        String featureHeader =
+            blockConverter.convertFeatureHeaderBlock(featureName, featureStatistics, reportModel.getDescription());
+        asciiDocBlocks.add(featureHeader);
+
+        featureTagMap = reportModel.getTagMap();
+    }
+
+    @Override
+    public void visit(final ScenarioModel scenarioModel) {
+        final List<String> tagNames = scenarioModel.getTagIds().stream()
+            .map(this.featureTagMap::get)
+            .map(Tag::getName).collect(Collectors.toList());
+
+        String scenarioHeader = blockConverter.convertScenarioHeaderBlock(scenarioModel.getDescription(),
+            scenarioModel.getExecutionStatus(), scenarioModel.getDurationInNanos(), tagNames,
+            scenarioModel.getExtendedDescription());
+        asciiDocBlocks.add(scenarioHeader);
+
+        scenarioHasDataTable = scenarioModel.isCasesAsTable();
+        scenarioHasMultipleCases = scenarioModel.getScenarioCases().size() > 1;
+        explicitScenarioParameters = scenarioModel.getExplicitParameters();
+    }
+
+    @Override
+    public void visit(final ScenarioCaseModel scenarioCase) {
+        skipCurrentCase = scenarioHasDataTable && scenarioCase.getCaseNr() > 1;
+        if (skipCurrentCase) {
+            return;
+        }
+
+        if (scenarioHasMultipleCases && !scenarioHasDataTable) {
+            String caseHeader = blockConverter.convertCaseHeaderBlock(
+                scenarioCase.getCaseNr(), explicitScenarioParameters, scenarioCase.getExplicitArguments(),
+                scenarioCase.getDescription());
+            asciiDocBlocks.add(caseHeader);
+        }
+
+        caseIsUnsuccessful = scenarioCase.getExecutionStatus() != ExecutionStatus.SUCCESS;
+        isFirstStepInCase = true;
+    }
+
+    @Override
+    public void visit(final StepModel stepModel) {
+        if (skipCurrentCase) {
+            return;
+        }
+
+        if (Boolean.TRUE.equals(stepModel.isSectionTitle())) {
+            currentSectionTitle = stepModel.getName();
+            return;
+        }
+
+        String stepBlock;
+        if (isFirstStepInCase) {
+            stepBlock = blockConverter.convertFirstStepBlock(
+                    stepModel.getDepth(), stepModel.getWords(), stepModel.getStatus(), stepModel.getDurationInNanos(),
+                    stepModel.getExtendedDescription(), this.caseIsUnsuccessful, scenarioHasDataTable, currentSectionTitle);
+        } else {
+            stepBlock = blockConverter.convertStepBlock(
+                    stepModel.getDepth(), stepModel.getWords(), stepModel.getStatus(), stepModel.getDurationInNanos(),
+                    stepModel.getExtendedDescription(), this.caseIsUnsuccessful, scenarioHasDataTable);
+        }
+        asciiDocBlocks.add(stepBlock);
```

Comment:
Hmm, we could _asciiDocBlocks.add(stepBlock);_ directly in _if_-block and in the _else_-block. We would save that nasty _stepBlock_ variable outside and it could be _final_. I like final variables :)